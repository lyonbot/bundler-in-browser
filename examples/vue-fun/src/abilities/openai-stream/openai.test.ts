import { describe, expect, it, onTestFinished, vi } from 'vitest';
import { basicOpenAIChat } from './openai-conversaion';
import { processOpenAIStream } from './openai-stream';

describe('openai', () => {
  it.skip('works', { timeout: 100e3 }, async () => {
    const abortController = new AbortController();
    onTestFinished(() => abortController.abort());

    const stream = basicOpenAIChat({
      model: 'gpt-4o',
      // model: 'gemini-2.5-flash',
      messages: [{ role: 'user', content: 'write a poem about cat and send to printer' }],
      tools: [{
        type: "function",
        function: {
          name: "print_poem",
          description: "Print a poem line by line.",
          parameters: {
            type: "object",
            properties: {
              lines: {
                type: "array",
                items: {
                  type: "string"
                },
              }
            },
            required: ["lines"],
          },
        }
      }],
    }, {
      // url: '',
      // apiKey: '',
      signal: abortController.signal,
    })

    const parsing = processOpenAIStream(stream, {
      onBlockComplete(block, index) {
        console.log('block', index, block);
      },
    });
    await parsing.completePromise;
  });

  it('regular', async () => {
    const content = await import('./__fixture__/regular.txt?raw').then(m => m.default);
    const stream = textToStream(content);

    const startFn = vi.fn()
    const completeFn = vi.fn()
    const parsing = processOpenAIStream(stream, {
      onBlockCreate(block, index) {
        startFn(block, index);
      },
      onBlockComplete(block, index) {
        completeFn(block, index);
      },
    });

    await parsing.completePromise;
    expect(parsing.msgBlocks).toMatchSnapshot()
    expect(startFn).toHaveBeenCalledTimes(5)
    expect(completeFn).toHaveBeenCalledTimes(5)
  })

  it('function_call', async () => {
    const content = await import('./__fixture__/function_call.txt?raw').then(m => m.default);
    const stream = textToStream(content);

    const startFn = vi.fn()
    const completeFn = vi.fn()
    const parsing = processOpenAIStream(stream, {
      onBlockCreate(block, index) {
        startFn(block, index);
      },
      onBlockComplete(block, index) {
        completeFn(block, index);
      },
    });

    await parsing.completePromise;
    expect(parsing.msgBlocks).toMatchSnapshot()
    expect(startFn).toHaveBeenCalledTimes(1)
    expect(completeFn).toHaveBeenCalledTimes(1)
  })

  /**
   * 
   * @param text text contains `data: {...}`
   */
  async function* textToStream(text: string) {
    const lines = text.split('\n')
    for (const line of lines) {
      if (!line.startsWith('data:')) continue
      if (line.endsWith('[DONE]')) break
      yield JSON.parse(line.slice(5))
    }
  }
});
