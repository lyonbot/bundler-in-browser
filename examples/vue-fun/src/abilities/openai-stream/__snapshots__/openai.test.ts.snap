// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openai > function_call 1`] = `
[
  {
    "completed": true,
    "content": {
      "function": {
        "arguments": "{"lines":["In a sunlit room, a cat does lie,","With a gentle purr and a sleepy eye.","Soft fur gleams in the afternoon light,","A creature of grace, a pure delight.","","Whiskers twitch in a playful dream,","Chasing shadows in a moonlit beam.","Paws tucked close, a curled-up ball,","In slumber deep, heed nature's call.","","A leap, a bound, through gardens fair,","With stealthy grace, the cat is there.","A hunter's heart in a playful chase,","In every move, a hint of grace.","","Eyes like emeralds, bright and wise,","Reflecting the world, capturing skies.","A feline friend, in joy and jest,","In quiet moments, they find their rest.","","Oh, gentle cat, with spirit free,","In your presence, peace and glee.","A purr, a meow, a soft embrace,","In your company, we find our space."]}",
        "name": "print_poem",
      },
      "id": "call_lCRDgpJssgwhEOV7k2Dkg2lv",
      "index": 0,
    },
    "type": "function",
  },
]
`;

exports[`openai > regular 1`] = `
[
  {
    "completed": true,
    "content": "Here's the classic "Hello World" program in Python:

",
    "type": "text",
  },
  {
    "completed": true,
    "content": "print("Hello, World!")
",
    "language": "python",
    "type": "code",
  },
  {
    "completed": true,
    "content": "### How to Run It:

1.  **Save the code:** Open a text editor (like Notepad on Windows, TextEdit on Mac, or any code editor like VS Code, Sublime Text, Atom). Copy and paste the line \`print("Hello, World!")\` into the editor. Save the file with a \`.py\` extension, for example, \`hello.py\`.

2.  **Open your terminal/command prompt:**
    *   **Windows:** Search for "cmd" or "PowerShell" in the Start menu.
    *   **Mac/Linux:** Search for "Terminal" in your applications.

3.  **Navigate to the directory:** Use the \`cd\` command to go to the directory where you saved your \`hello.py\` file. For example, if you saved it in a folder called \`my_python_programs\` on your desktop, you might type:
    \`\`\`bash
    cd Desktop/my_python_programs
    \`\`\`

4.  **Run the program:** Once you're in the correct directory, type:
    \`\`\`bash
    python hello.py
    \`\`\`
    (On some systems, you might need to use \`python3\` instead of \`python\` if you have multiple Python versions installed).

### Expected Output:

",
    "type": "text",
  },
  {
    "completed": true,
    "content": "Hello, World!
",
    "language": "",
    "type": "code",
  },
  {
    "completed": true,
    "content": "",
    "type": "text",
  },
]
`;
